test/e2e/cache/cache_server_test.go

```diff
@@ -0,0 +1,504 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"path"
+	"strconv"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/kcp-dev/logicalcluster/v2"
+	"github.com/stretchr/testify/require"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	apimachineryerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	genericrequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/client-go/dynamic"
+	kubernetesscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+
+	cacheclient "github.com/kcp-dev/kcp/pkg/cache/client"
+	cacheserver "github.com/kcp-dev/kcp/pkg/cache/server"
+	cacheopitons "github.com/kcp-dev/kcp/pkg/cache/server/options"
+	"github.com/kcp-dev/kcp/pkg/embeddedetcd"
+	"github.com/kcp-dev/kcp/test/e2e/framework"
+)
+
+type testScenario struct {
+	name string
+	work func(ctx context.Context, t *testing.T, cacheClientRT *rest.Config, cluster logicalcluster.Name, gvr schema.GroupVersionResource)
+}
+
+// scenarios holds all test scenarios
+var scenarios = []testScenario{
+	{"TestSchemaIsNotEnforced", testSchemaIsNotEnforced},
```

Comment:
my comment on the names is outstanding - the godoc you have on the methods would be good to have in the test name, when I break this functionality I want to understand what I'm looking at :)

Reply:
I think Steve is saying that we want real test funcs like `TestSchemaIsNotEnforced` instead of some strings here with that name. I.e. put the tests at the top, and the mechanics to run them into helpers.

Reply:
we could do that in the future when we have a function (in the framework) that knows how to create a shared instance of the cache server. Before that the tests have to be called via `TestCacheServerAllScenarios` method. 

I'll move the `scenarios` struct closer to the the `TestCacheServerAllScenarios`.